tpp-app/src/services/AppleService.js

```diff
@@ -0,0 +1,135 @@
+        [APPLE_CREDENTIALS.GIVEN_NAME, JSON.stringify(givenName)],
+        [APPLE_CREDENTIALS.FAMILY_NAME, JSON.stringify(familyName)],
+        [APPLE_CREDENTIALS.IDENTITY_TOKEN, JSON.stringify(token)]
+        ])
+        .then(() => {
+            console.log(`Saved user's Apple identity as: ${userId}, ${givenName}, ${familyName}, ${token}`);
+            resolve();
+            })
+        .catch((e) => {
+            console.log(`POSTAppleIdentity error: ${JSON.stringify(e)}`);
+            reject();
+        });
+});
+
+/**
+ * Retrieve the user's identity token so we can authenticate them
+ */
+export const GETAppleIdentityToken = async () => new Promise( async (resolve, reject) => {
+    AsyncStorage.getItem(APPLE_CREDENTIALS.IDENTITY_TOKEN)
+        .then((token) => {
+            console.log("Retrieved user's Apple identity token");
+            resolve(JSON.parse(token));
+        }).catch((e) => {
+            console.log(`GETAppleIdentityToken error: ${JSON.stringify(e)}`);
+            reject();
+    });
+});
+
+/**
+ * Retrieve the user's user id so we can authenticate them
+ */
+export const GETAppleUser = async () => new Promise( async (resolve, reject) => {
+    AsyncStorage.getItem(APPLE_CREDENTIALS.USER_ID)
+        .then((token) => {
+            console.log("Retrieved user's Apple user ID", token);
+            resolve(JSON.parse(token));
+        }).catch((e) => {
+        console.log(`GETAppleUser error: ${JSON.stringify(e)}`);
+        reject();
+    });
+});
+
+/**
+ * Retrieve last saved time
+ */
+export const GETLastSavedBackupTime = async () => new Promise( async (resolve, reject) => {
+    AsyncStorage.getItem(ICLOUD.LAST_SAVED_TIME)
+        .then((lastSavedTime) => {
+            console.log("Retrieved user's last back up time");
+            resolve(JSON.parse(lastSavedTime));
+        }).catch((e) => {
+        console.log(`GETLastSavedBackupTime error: ${JSON.stringify(e)}`);
+        reject();
+    });
+});
+
+/**
+ * Backup account to iCloud
+ */
+export const POSTBackupToiCloud = async () => new Promise( async (resolve, reject) => {
+    try {
+        // Create a json of everything in AsyncStorage
+        const keys = await AsyncStorage.getAllKeys();
+        const values = await AsyncStorage.multiGet(keys);
+
+        // Save keys and values to iCloud
+        await iCloudStorage.setItem(ICLOUD.BACKUP_KEY, JSON.stringify(values));
+        await AsyncStorage.setItem(ICLOUD.LAST_SAVED_TIME, JSON.stringify(new Date));
+
+        resolve();
+    } catch (e) {
+        console.log(`POSTBackupToiCloud error: ${JSON.stringify(e)}`);
+        reject();
+    }
+});
+
+/**
+ * Check if iCloud has backed up data
+ * @returns boolean indicating whether the user has a backup of M. Nation data in their iCloud
+ */
+export const userHasiCloudBackup = async () => new Promise(async(resolve, reject) => {
+    try {
+        const iCloudKeys = await iCloudStorage.getAllKeys();
```

Comment:
idk if time is an issue but wouldn't just calling `iCloudStorage.getItem(ICLOUD.BACKUP_KEY)` and null checking it be faster? Rather than needing to get all keys. Idk how this API is implemented though

Reply:
nvm i think this is right since you'd be retrieving the whole backup if the key exists. Think that's more expensive than getting all the keys

Reply:
that's what I thought too but I'll look it up to see if there's any docs. Because if their backup is huuuuuge ```iCloudStorage.getItem(ICLOUD.BACKUP_KEY)``` could end up being more costly than ```iCloudStorage.getAllKeys();```

Reply:
since the react-native-icloudstorage is like a wrapper copying AsyncStorage that syncs with iCloud, there should only ever be zero or one keys in it for our app's container which is the backup key.